docs/guides/custom-functions.md

```diff
@@ -0,0 +1,131 @@
+# Custom Functions
+
+If the built-in functions are not enough, Spectral allows you to write and use your own custom functions.
+
+A custom function might be any JS function compliant with [IFunction](https://github.com/stoplightio/spectral/blob/90a0864863fa232bf367a26dace61fd9f93198db/src/types/function.ts#L3#L8) type.
+
+```ts
+export type IFunction<O = any> = (
+  targetValue: any,
+  options: O,
+  paths: IFunctionPaths,
+  otherValues: IFunctionValues,
+) => void | IFunctionResult[];
+```
+
+Custom functions take exactly the same arguments as built-in functions do, so you are more than welcome to take a look at the existing implementation.
+
+The process of creating a function involves 2 steps:
+
+- create a js file inside of a directory called `functions` that should be placed next to your ruleset file
+- create a `functions` array in your ruleset if you haven't done it yet and place a string with the filename without `.js` extension
+
+#### Example:
+
+**functions/abc.js**
+
+```js
+module.exports = (targetVal) => {
+  if (targetVal !== 'abc') {
+    return [
+      {
+        message: 'Value must equal "abc".',
+      },
+    ];
+  }
+};
+```
+
+**my-ruleset.yaml**
+
+```yaml
+functions: [abc]
+rules:
+  my-rule:
+    message: "{{error}}"
+    given: "$.info"
+    then:
+      function: "abc"
+```
+
+Optionally, if you'd like to validate the data that it passed to abc function before the function gets executed, you can provide a JSON Schema.
```

Comment:
Again, I'm not sure that this is what it should be doing?

We have a json schema rule that should be used to validate target data.

This JSON schema should describe the `functionOptions` available in this function. Eventually we'll use it to render documentation for the ruleset and custom functions, and I guess we could also use it to validate the `functionOptions` the end user passes in?

Reply:
> This JSON schema should describe the functionOptions available in this function. Eventually we'll use it to render documentation for the ruleset and custom functions, and I guess we could also use it to validate the functionOptions the end user passes in?

Yeah, we validate the function options that are passed in.